gazelle_cabal/dependency_resolution.go

```diff
@@ -223,85 +214,74 @@ func getPackageLabel(
 	pkgName string,
 	from label.Label,
 ) label.Label {
-	var resolvers []func(*resolve.RuleIndex, string, string, label.Label)(label.Label, error)
-	resolvers = append(resolvers,
-		resolveAgainstPrivateLibraries,
-		resolveAgainstPublicLibraries,
-		resolveAgainstRepository,
-	)
-
-	for _, resolver := range resolvers {
-		lbl, err := resolver(ix, packageRepo, pkgName, from)
-		if (err == nil) {
-			return lbl
+	for _, searchScope := range obtainSearchScopes(pkgName, from) {
+		if labelFound, err := searchInLibraries(ix, packageRepo, pkgName, from, searchScope); err == nil {
+			return labelFound
 		}
 	}
 
-	log.Fatalf("Could not resolve dependency '%s' from '%s'", pkgName, from)
-	return label.Label{}
+	return grabFromRepository(packageRepo, pkgName, from)
 }
 
-func resolveAgainstPrivateLibraries(
-	ix *resolve.RuleIndex,
-	packageRepo string,
-	pkgName string,
-	from label.Label,
-) (label.Label, error) {
-	parentPackage, depName := splitDependencyString(pkgName)
-
-	if len(parentPackage) > 0 && (parentPackage != depName) {
-		return label.Label{}, fmt.Errorf("It looks like %s is not an internal, private library", pkgName)
-	}
-
-	// Search for the rule of an internal library using the prefix "internal_library:" for the key
-	spec := resolve.ImportSpec{gazelleCabalName, "private_library:" + depName}
-	res := ix.FindRulesByImport(spec, gazelleCabalName)
-
-	// Search for the label of an internal library in the current package
-	for _, r := range res {
-		if r.IsSelfImport(from) {
-			log.Fatalf("Dependency cycle detected in the following component: %s", from)
-		}
-
-		// indeed internal library can be refereced form the package of its definition
-		if (r.Label.Repo == from.Repo && r.Label.Pkg == from.Pkg) {
-			return rel(r.Label, from), nil
+// Dependency string may contain 'experimental' colon syntax e.g package-b:sublib
+func obtainSearchScopes(depName string, from label.Label) []string {
+	splitted := strings.Split(depName, ":")
+	publicPrefix := "public_library"
+	privatePrefix := "private_library"
+	format := "%s:%s:%s"
+	if len(splitted) > 1 {
+		cabalPackageId := splitted[0]
+		libraryId      := splitted[1]
+		return []string{
+			// there is prefix which can reference the public or internal library
+	        fmt.Sprintf(format, publicPrefix, cabalPackageId, libraryId),
+			// TODO: potentially throw error if from.Pkg != cabalPackageId?
+			fmt.Sprintf(format, privatePrefix, from.Pkg, libraryId),
 		}
 	}
 
-	return label.Label{}, fmt.Errorf("Internal library '%s' not found", pkgName)
+	return []string{
+		// it can be localally defined library which can be either private or public
+		fmt.Sprintf(format, privatePrefix, from.Pkg, depName),
+		fmt.Sprintf(format, publicPrefix, from.Pkg, depName),
```

Comment:
I think every time this function writes `from.Pkg`, it is really meant `r.PrivateAttr("pkgName")` where `r` is the rule parameter from `setDepsAndPluginsAttributes`.

The package of a Bazel label is not usually a valid Haskell package.

Reply:
Ok, you're right. Will fix that.